Add inline symbol reference count hints

gosvig123 opened the pull request:
## Summary

• Adds inline reference count hints that display next to symbols in the editor, showing how many times each symbol is referenced throughout the codebase
• Integrates with existing inlay hints system and LSP to provide accurate, real-time reference counts with proper debouncing and multi-buffer handling

This feature helps developers quickly understand code usage patterns and make informed refactoring decisions by showing reference counts directly in the editor, similar to VSCode's reference count feature.

## Release Notes:

- Added: Inline symbol reference count hints that display next to symbols showing how many times they are referenced throughout the codebase (toggle with `editor: toggle symbol ref hints`)


Screenshots below 
<img width="1095" height="616" alt="Screenshot 2025-09-29 at 20 17 15" src="https://github.com/user-attachments/assets/4d3bc7cb-948f-4c51-bf51-c1b64d8a5016" />
<img width="1190" height="553" alt="Screenshot 2025-09-29 at 20 17 06" src="https://github.com/user-attachments/assets/421f9769-be0a-4414-995b-51d1d07db176" />
<img width="998" height="373" alt="Screenshot 2025-09-29 at 20 16 40" src="https://github.com/user-attachments/assets/99d5a680-00df-4a79-8ce2-55c2c4723d1c" />
<img width="758" height="237" alt="Screenshot 2025-09-29 at 20 15 47" src="https://github.com/user-attachments/assets/b93684f5-56a1-4831-99d2-f8ebb760cd01" />

gosvig123 committed:
mvp of symbol inline count

gosvig123 committed:
working version pending multibuffer sorting and cleaning

gosvig123 committed:
disabled for multi buffer and created relationship to inlay hints

gosvig123 committed:
fixed refresh and multibuffer stabillity,  but pending code clean for duplicate functionality

gosvig123 committed:
working version with cleaned logs, pendings all language server functionality

gosvig123 committed:
making it work with different lsps

gosvig123 committed:
reduced duplicate logic into helpers

gosvig123 committed:
fixed linting

gosvig123 committed:
lint errors

gosvig123 committed:
spelling

gosvig123 committed:
added fix for cancelation test

gosvig123 committed:
fixed formatting

SomeoneToIgnore reviewed (commented):
Can you attach the screenshots of the change, [please](https://github.com/zed-industries/zed/blob/main/CONTRIBUTING.md#sending-changes)?

gosvig123 commented:
> Can you attach the screenshots of the change, [please](https://github.com/zed-industries/zed/blob/main/CONTRIBUTING.md#sending-changes)?

Yea of course i added screenshots above from before, after and with different themes and languages, also added one with the toggle in use just in case.

Just to be clear i expect this to default to false although it is true right now, don't want to impose a version of codelens but do really want i badly and think this provides for a nice base that is abstracted and can bed edited as the team sees fit, in this amazing editor :)

SomeoneToIgnore reviewed (changes requested):
I have compiled a branch and run it on https://github.com/rust-lang/rust-analyzer/ : 
https://github.com/user-attachments/assets/bab33632-0173-4382-967b-41a49db8249d

I'm quite confused by what I see: 

* new inlays appear without any configuration and toggle action

> Just to be clear i expect this to default to false although it is true right now

sounds great, but why does not it work?

I also think that there's no point to have an action for such thing, but rather a config option under inlay hints (disabled by default for sure).

* there's A LOT of the new inlays, in each place — this is quite a visual mess, and all of them are clearly wrong (there's more than 0 usages of the items for sure)

This is very noisy and clearly does not help now.

* every time I scroll, we seem to re-request the hints?

This is insanely expensive and scrolling the editor is a pain in this branch's build, it stutters and looses frames quite badly — admittedly, `main` does not feel 100% smooth either, but here it's very bad.

This is also quite resource-demanding for language server and may be very long (in case of r-a, for instance).

Caching this all seems like a large task on itself, as e.g. editing the buffer will invalidate this.

* the design is also debatable

Before this is merged, I'd have to call people from design team to assert that, but based in my experience, my PRs were called out and reverted with a smaller visual oddities — and this version won't pass: it's too many symbols blending into the text.

I'm also concerned that various code parts like enums, which were initially lined by the left part, will "dance" due to different, real digits (non-0) shown for each enum variant as a usage counter.
This will look not that nice, I anticipate?

* there are no tests at all....

I did not look at the code in too much detail, but I see that it's rigged with odd comments like

<img width="303" height="84" alt="Image" src="https://github.com/user-attachments/assets/42863ea5-ca42-445c-bfc3-5a746cd41a8d" />

and hacks like debuger hints' ids used.

-----------------------

I went on and tested on a random C file with the same confusing result — that leaves me very suspicious we'll never merge this PR in its current form.

I expect it would take me a week or two of work (knowing what to do) to implement this — and this PR is far from working normally, so I'll close this PR as otherwise it will hang in the queue for quite some time.

If I was working on this, I'd concentrate on the next 2 large tasks:

* reconsider the design to show things similar to tasks, maybe?
In a gutter, or, if inlays are needed, consider showing them only on the lines where selections are? 
Or, not the lines but the blocks?
r-a in VSCode uses code lens, but those are not very well received as a design element in Zed. How other editors do this? 

Maybe, another way to avoid those tables of digits? 

* add tests, deal with caching and call as less such requests as possible, as those might be expensive.
Also, because the current branch drops frames, it means the new code has a bug where the main thread is occupied by something heavy — that has to be discovered and fixed for a real version
